core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java

```diff
@@ -0,0 +1,334 @@
+                }
+
+                // find any manifests that are no longer needed
+                try (CloseableIterable<ManifestFile> manifests = readManifestFiles(snapshot)) {
+                  for (ManifestFile manifest : manifests) {
+                    if (!validManifests.contains(manifest.path())) {
+                      manifestsToDelete.add(manifest.path());
+
+                      boolean isFromAncestor = ancestorIds.contains(manifest.snapshotId());
+                      boolean isFromExpiringSnapshot = expiredIds.contains(manifest.snapshotId());
+
+                      if (isFromAncestor && manifest.hasDeletedFiles()) {
+                        // Only delete data files that were deleted in by an expired snapshot if
+                        // that napshot is an ancestor of the current table state. Otherwise, a
+                        // snapshot
+                        // that deleted files and was rolled back will delete files that could be in
+                        // the current
+                        // table state.
+                        manifestsToScan.add(manifest.copy());
+                      }
+
+                      if (!isFromAncestor && isFromExpiringSnapshot && manifest.hasAddedFiles()) {
+                        // Because the manifest was written by a snapshot that is not an ancestor of
+                        // the current table state, the files added in this manifest can be removed.
+                        // The
+                        // extra check whether the manifest was written by a known snapshot that was
+                        // expired in this commit ensures that the full ancestor list between when
+                        // the snapshot
+                        // was written and this expiration is known and there is no missing history.
+                        // If
+                        // history were missing, then the snapshot could be an ancestor of the table
+                        // state
+                        // but the ancestor ID set would not contain it and this would be unsafe.
+                        manifestsToRevert.add(manifest.copy());
+                      }
+                    }
+                  }
+                } catch (IOException e) {
+                  throw new RuntimeIOException(
+                      e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
+                }
+
+                // add the manifest list to the delete set, if present
+                if (snapshot.manifestListLocation() != null) {
+                  manifestListsToDelete.add(snapshot.manifestListLocation());
+                }
+              }
+            });
+
+    deleteDataFiles(manifestsToScan, manifestsToRevert, validIds, afterExpiration);
+    LOG.warn("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
+    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
```

Comment:
@rdblue This warn logging is in the current implementation of file cleanup https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/RemoveSnapshots.java#L538

I've removed it from the reachable file cleanup implementation but should I remove it from here as well? I'm not sure if warn logging was done here intentionally but it also isn't really a warning level operation because we actually want to delete files in this operation.

Reply:
No, let's leave it in if it was there originally. Thanks for noting it!